In [ ]:
# importing the libraries
import numpy as np
import pandas as pd

### 2.6 Pandas -  Series and DataFrames

**Pandas Series**
* Pandas Series is a one-dimensional labeled array/list capable of holding data of any type (integer, string, float, python objects, etc.). 
* The labels are collectively called index. 
* Pandas Series can be thought as a single column of an excel spreadsheet and each entry in a series corresponds to an individual row in the spreadsheet.

In [ ]:
# creating a list of price of different medicines
med_price_list = [55,25,75,40,90]

# converting the med_price_list to an array 
med_price_arr = np.array(med_price_list)

# converting the list and array into a Pandas Series object  
series_list = pd.Series(med_price_list)
series_arr = pd.Series(med_price_arr)

# printing the converted series object
print(series_list)
print(series_arr)

0    55
1    25
2    75
3    40
4    90
dtype: int64
0    55
1    25
2    75
3    40
4    90
dtype: int64


* We can see that the list and array have been converted to a Pandas Series object.
* We also see that the series has automatically got index labels. Let's see how these can be modified.

In [ ]:
# changing the index of a series
med_price_list_labeled = pd.Series(med_price_list, index = ['Omeprazole','Azithromycin','Metformin','Ibuprofen','Cetirizine'])
print(med_price_list_labeled)

Omeprazole      55
Azithromycin    25
Metformin       75
Ibuprofen       40
Cetirizine      90
dtype: int64


**Performing mathematical operations on Pandas Series**

* The price of each medicine was increased by $2.5. Let's add this to the existing price.

In [ ]:
# adding 2.5 to existing prices
med_price_list_labeled_updated = med_price_list_labeled + 2.5
med_price_list_labeled_updated

Omeprazole      57.5
Azithromycin    27.5
Metformin       77.5
Ibuprofen       42.5
Cetirizine      92.5
dtype: float64

* A new price list was released by vendors for each medicine. Let's find the difference between new price and the old price

In [ ]:
new_price_list = [77, 45.5, 100, 50, 80]
new_price_list_labeled = pd.Series(new_price_list, index = ['Omeprazole','Azithromycin','Metformin','Ibuprofen','Cetirizine'])
print(new_price_list_labeled)

Omeprazole       77.0
Azithromycin     45.5
Metformin       100.0
Ibuprofen        50.0
Cetirizine       80.0
dtype: float64


In [ ]:
print('Difference between new price and old price - ')
print(new_price_list_labeled - med_price_list_labeled_updated)

Difference between new price and old price - 
Omeprazole      19.5
Azithromycin    18.0
Metformin       22.5
Ibuprofen        7.5
Cetirizine     -12.5
dtype: float64


**Pandas DataFrame**

Pandas DataFrame is a two-dimensional tabular data structure with labeled axes (rows and columns).

**Creating a Pandas DataFrame using a list**

In [ ]:
student = ['Mary', 'Peter', 'Susan', 'Toby', 'Vishal']
df1 = pd.DataFrame(student,columns=['Student'])
df1

Student
0    Mary
1   Peter
2   Susan
3    Toby
4  Vishal

**Creating a Pandas DataFrame using a dictionary**

In [ ]:
# defining another list
grades = ['B-','A+','A-', 'B+', 'C']

# creating the dataframe using a dictionary
df2 = pd.DataFrame({'Student':student,'Grade':grades})
df2

Student Grade
0    Mary    B-
1   Peter    A+
2   Susan    A-
3    Toby    B+
4  Vishal     C

**Creating a Pandas DataFrame using Series**

The data for total energy consumption for the U.S. was collected from 2012 - 2018. Let's see how this data can be presented in form of data frame.

In [ ]:
year = pd.Series([2012,2013,2014,2015,2016,2017,2018])
energy_consumption = pd.Series([2152,2196,2217,2194,2172,2180,2258])

df3 = pd.DataFrame({'Year':year,'Energy_Consumption(Mtoe)':energy_consumption})
df3

Year  Energy_Consumption(Mtoe)
0  2012                      2152
1  2013                      2196
2  2014                      2217
3  2015                      2194
4  2016                      2172
5  2017                      2180
6  2018                      2258

**Creating a Pandas DataFrame using random values**

For encryption purposes a web browser company wants to generate random values which have mean equal to 0 and variance equal to 1. They want 5 randomly generated numbers in 2 different trials.

In [ ]:
# we can create a new dataframe using random values 
df4 = pd.DataFrame(np.random.randn(5,2),columns = ['Trial 1', 'Trial 2'])
df4

Trial 1   Trial 2
0  0.740967  0.935937
1  1.519597  0.013860
2  0.259383 -0.824822
3 -0.744735 -0.039417
4  0.156586 -0.348059

In [ ]:
np.random.randn(5,2)

array([[-0.57825811, -0.28591443],
       [-0.56854768,  0.89600388],
       [ 0.12066679,  1.42525467],
       [-1.29101572, -0.70963579],
       [-1.03733844, -0.54760756]])

### 2.7 Pandas - Accessing and Modifying

**Accessing Series**

The revenue (in billion dollars) of different telecommunication operators in U.S. was collected for the year of 2020. The following lists consist of the names of the telecommunication operators and their respective revenue (in billion dollars).

In [ ]:
operators = ['AT&T', 'Verizon', 'T-Mobile US', 'US Cellular']
revenue = [171.76, 128.29, 68.4, 4.04]

#creating a Series from lists
telecom = pd.Series(revenue, index=operators)
telecom

AT&T           171.76
Verizon        128.29
T-Mobile US     68.40
US Cellular      4.04
dtype: float64

**Accessing Pandas Series using its index**

In [ ]:
# accessing the first element of series 
telecom[0]

171.76

In [ ]:
#  accessing firt 3 elements of a series
telecom[:3]

AT&T           171.76
Verizon        128.29
T-Mobile US     68.40
dtype: float64

In [ ]:
# accessing the last two elements of a series
telecom[-2:]

T-Mobile US    68.40
US Cellular     4.04
dtype: float64

In [ ]:
# accessing multiple elements of a series
telecom[[0,2,3]]

AT&T           171.76
T-Mobile US     68.40
US Cellular      4.04
dtype: float64

**Accessing Pandas Series using its labeled index**

In [ ]:
# accessing the revenue of AT&T
telecom['AT&T']

171.76

In [ ]:
#  accessing firt 3 revenues of operators in the series
telecom[:'T-Mobile US']

AT&T           171.76
Verizon        128.29
T-Mobile US     68.40
dtype: float64

In [ ]:
# accessing multiple values
telecom[['AT&T','US Cellular','Verizon']]

AT&T           171.76
US Cellular      4.04
Verizon        128.29
dtype: float64

**Accessing DataFrames**

The data of the customers visiting 24/7 Stores from different locations was collected. The data includes Customer ID, location of store, gender of the customer,  type of product purchased, quantity of products purchased, total bill amount. Let's create the dataset and see how to access different entries of it.

In [ ]:
# creating the dataframe using dictionary
store_data = pd.DataFrame({'CustomerID': ['CustID00','CustID01','CustID02','CustID03','CustID04']
                           ,'location': ['Chicago', 'Boston', 'Seattle', 'San Francisco', 'Austin']
                           ,'gender': ['M','M','F','M','F']
                           ,'type': ['Electronics','Food&Beverages','Food&Beverages','Medicine','Beauty']
                           ,'quantity':[1,3,4,2,1],'total_bill':[100,75,125,50,80]})
store_data

CustomerID       location gender            type  quantity  total_bill
0   CustID00        Chicago      M     Electronics         1         100
1   CustID01         Boston      M  Food&Beverages         3          75
2   CustID02        Seattle      F  Food&Beverages         4         125
3   CustID03  San Francisco      M        Medicine         2          50
4   CustID04         Austin      F          Beauty         1          80

In [ ]:
data_customer = pd.DataFrame({'customer': ['101','102','103','104']
                           ,'category': ['cat2', 'cat2', 'cat1', 'cat3']
                           ,'important': ['yes','no','yes','yes']
                           ,'sales': [123,52,214,663]
                           })

data_customer['category'].value_counts()
data_customer['category'].unique()
data_customer['customer'].value_counts()
data_customer
data_customer.loc[[1,2], ['customer', 'sales']]
data_customer.loc[1:3, ['customer', 'sales']]

KeyError: ignored

In [ ]:
calorie_data = pd.DataFrame({'day': ['day1','day2','day3','day4','day5']
                           ,'calories': [450, 300, 345, 520, 600]
                           ,'duration_min': [30, 25, 29, 39, 48]
                           })
calorie_data
calorie_data[:2]

day  calories  duration_min
0  day1       450            30
1  day2       300            25

In [ ]:
# accessing first row of the dataframe
store_data[:1]

CustomerID location gender         type  quantity  total_bill
0   CustID00  Chicago      M  Electronics         1         100

In [ ]:
# accessing first column of the dataframe
store_data['location']

0          Chicago
1           Boston
2          Seattle
3    San Francisco
4           Austin
Name: location, dtype: object

In [ ]:
# accessing rows with the step size of 2
store_data[::2] 

CustomerID location gender            type  quantity  total_bill
0   CustID00  Chicago      M     Electronics         1         100
2   CustID02  Seattle      F  Food&Beverages         4         125
4   CustID04   Austin      F          Beauty         1          80

In [ ]:
# accessing the rows in reverse
store_data[::-2]

CustomerID location gender            type  quantity  total_bill
4   CustID04   Austin      F          Beauty         1          80
2   CustID02  Seattle      F  Food&Beverages         4         125
0   CustID00  Chicago      M     Electronics         1         100

**Using loc and iloc method**

**loc method**

* loc is a  method to access rows and columns on pandas objects. When using the loc method on a dataframe, we specify which rows and which columns we want by using the following format:

  * **dataframe.loc[row selection, column selection]**

* DataFrame.loc[] method is a method that takes **only index labels** and returns row or dataframe if the index label exists in the data frame.

In [ ]:
# accessing first index value using loc method (indexing starts from 0 in python)
store_data.loc[1]

CustomerID          CustID01
location              Boston
gender                     M
type          Food&Beverages
quantity                   3
total_bill                75
Name: 1, dtype: object

**Accessing selected rows and columns using loc method**

In [ ]:
# accessing 1st and 4th index values along with location and type columns 
store_data.loc[[1,4],['location','type']]

location            type
1   Boston  Food&Beverages
4   Austin          Beauty

**iloc method**

* The iloc indexer for Pandas Dataframe is used for **integer location-based** indexing/selection by position. When using the loc method on a dataframe, we specify which rows and which columns we want by using the following format:

  * **dataframe.iloc[row selection, column selection]**



In [ ]:
# accessing selected rows and columns using iloc method 
store_data.iloc[[1,4],[0,2]]

CustomerID gender
1   CustID01      M
4   CustID04      F

**Difference between loc and iloc indexing methods**

* loc is label-based, which means that you have to specify rows and columns based on their row and column labels.
* iloc is integer position-based, so you have to specify rows and columns by their integer position values (0-based integer position).


If we use labels instead of index values in .iloc it will throw an error.

In [ ]:
# accessing selected rows and columns using iloc method 
store_data.iloc[[1,4],['location','type']]

IndexError: ignored

* As expected, .iloc has given error on using 'labels'.

We can modify entries of a dataframe using loc or iloc too

In [ ]:
print(store_data.loc[4,'type'])
store_data.loc[4,'type'] = 'Electronics'

Beauty


In [ ]:
store_data

CustomerID       location gender            type  quantity  total_bill
0   CustID00        Chicago      M     Electronics         1         100
1   CustID01         Boston      M  Food&Beverages         3          75
2   CustID02        Seattle      F  Food&Beverages         4         125
3   CustID03  San Francisco      M        Medicine         2          50
4   CustID04         Austin      F     Electronics         1          80

In [ ]:
store_data.iloc[4,3] = 'Beauty'
store_data

CustomerID       location gender            type  quantity  total_bill
0   CustID00        Chicago      M     Electronics         1         100
1   CustID01         Boston      M  Food&Beverages         3          75
2   CustID02        Seattle      F  Food&Beverages         4         125
3   CustID03  San Francisco      M        Medicine         2          50
4   CustID04         Austin      F          Beauty         1          80

In [ ]:
calorie_data = pd.DataFrame({'day': ['day1','day2','day3','day4','day5']
                           ,'calories': [450, 300, 345, 520, 600]
                           ,'duration_min': [30, 25, 29, 39, 48]
                           })
calorie_data

day  calories  duration_min
0  day1       450            30
1  day2       300            25
2  day3       345            29
3  day4       520            39
4  day5       600            48

**Condition based indexing**

In [ ]:
store_data['quantity']>1

0    False
1     True
2     True
3     True
4    False
Name: quantity, dtype: bool

* Wherever the condition of greater than 1 is satisfied in quantity column, 'True' is returned. Let's retrieve the original values wherever the condition is satisfied.

In [ ]:
store_data.loc[store_data['quantity']>1]

CustomerID       location gender            type  quantity  total_bill
1   CustID01         Boston      M  Food&Beverages         3          75
2   CustID02        Seattle      F  Food&Beverages         4         125
3   CustID03  San Francisco      M        Medicine         2          50

In [ ]:
calorie_data = pd.DataFrame({'day': ['day1','day2','day3','day4','day5']
                           ,'calories': [450, 300, 345, 520, 600]
                           ,'duration_min': [30, 25, 29, 39, 48]
                           })
calorie_data

day  calories  duration_min
0  day1       450            30
1  day2       300            25
2  day3       345            29
3  day4       520            39
4  day5       600            48

* Wherever the condition is satisfied we get the original values, and wherever the condition is not satisfied we do not get those records in the output.

**Column addition and removal from a Pandas DataFrame**

**Adding a new column in a DataFrame**

In [ ]:
store_data

CustomerID       location gender            type  quantity  total_bill
0   CustID00        Chicago      M     Electronics         1         100
1   CustID01         Boston      M  Food&Beverages         3          75
2   CustID02        Seattle      F  Food&Beverages         4         125
3   CustID03  San Francisco      M        Medicine         2          50
4   CustID04         Austin      F          Beauty         1          80

In [ ]:
# adding a new column in data frame store_data which is a rating (out of 5) given by customer based on their shopping experience
store_data['rating'] = [2,5,3,4,4]
store_data

CustomerID       location gender  ... quantity  total_bill  rating
0   CustID00        Chicago      M  ...        1         100       2
1   CustID01         Boston      M  ...        3          75       5
2   CustID02        Seattle      F  ...        4         125       3
3   CustID03  San Francisco      M  ...        2          50       4
4   CustID04         Austin      F  ...        1          80       4

[5 rows x 7 columns]

**Removing a column from a DataFrame**

* The CustomerID column is a unique identifier of each customer. This unique identifier will not help 24/7 Stores in getting useful insights about their customers. So, they have decided to remove this column from the data frame.

In [ ]:
store_data.drop('CustomerID',axis=1)

location gender            type  quantity  total_bill  rating
0        Chicago      M     Electronics         1         100       2
1         Boston      M  Food&Beverages         3          75       5
2        Seattle      F  Food&Beverages         4         125       3
3  San Francisco      M        Medicine         2          50       4
4         Austin      F          Beauty         1          80       4

* We sucessfully removed the 'CustomerID' from dataframe. But this change is not permanent in the dataframe, let's have a look at the store_data again.

In [ ]:
store_data

CustomerID       location gender  ... quantity  total_bill  rating
0   CustID00        Chicago      M  ...        1         100       2
1   CustID01         Boston      M  ...        3          75       5
2   CustID02        Seattle      F  ...        4         125       3
3   CustID03  San Francisco      M  ...        2          50       4
4   CustID04         Austin      F  ...        1          80       4

[5 rows x 7 columns]

* We see that store_data still has column 'CustomerID' in it. 
* To make permanent changes to a dataframe there are two methods will have to use a parameter `inplace` and set its value to `True`.

In [ ]:
store_data.drop('CustomerID',axis=1,inplace=True)
store_data

location gender            type  quantity  total_bill  rating
0        Chicago      M     Electronics         1         100       2
1         Boston      M  Food&Beverages         3          75       5
2        Seattle      F  Food&Beverages         4         125       3
3  San Francisco      M        Medicine         2          50       4
4         Austin      F          Beauty         1          80       4

* Now the column has been permanently removed from the dataframe.

In [ ]:
# we can also remove multiple columns simultaneously 
# it is always a good idea to store the new/updated data frames in new variables to avoid changes to the existing data frame

# creating a copy of the existing data frame
new_store_data = store_data.copy()
store_data

location gender            type  quantity  total_bill  rating
0        Chicago      M     Electronics         1         100       2
1         Boston      M  Food&Beverages         3          75       5
2        Seattle      F  Food&Beverages         4         125       3
3  San Francisco      M        Medicine         2          50       4
4         Austin      F          Beauty         1          80       4

In [ ]:
# dropping location and rating columns simultaneously
new_store_data.drop(['location','rating'],axis=1,inplace=True)
new_store_data

gender            type  quantity  total_bill
0      M     Electronics         1         100
1      M  Food&Beverages         3          75
2      F  Food&Beverages         4         125
3      M        Medicine         2          50
4      F          Beauty         1          80

In [ ]:
# lets check if store_data was impacted
store_data

location gender            type  quantity  total_bill  rating
0        Chicago      M     Electronics         1         100       2
1         Boston      M  Food&Beverages         3          75       5
2        Seattle      F  Food&Beverages         4         125       3
3  San Francisco      M        Medicine         2          50       4
4         Austin      F          Beauty         1          80       4

* There were no changes to data frame store_data.

* Deep copy stores copies of the object’s value.
* Shallow Copy stores the references of objects to the original memory address. 

**Removing rows from a dataframe**

In [ ]:
store_data.drop(1,axis=0)

location gender            type  quantity  total_bill  rating
0        Chicago      M     Electronics         1         100       2
2        Seattle      F  Food&Beverages         4         125       3
3  San Francisco      M        Medicine         2          50       4
4         Austin      F          Beauty         1          80       4

In [ ]:
store_data

location gender            type  quantity  total_bill  rating
0        Chicago      M     Electronics         1         100       2
1         Boston      M  Food&Beverages         3          75       5
2        Seattle      F  Food&Beverages         4         125       3
3  San Francisco      M        Medicine         2          50       4
4         Austin      F          Beauty         1          80       4

* Notice that we used **`axis=0`** to drop a row from a data frame, while we were using **`axis=1`** for dropping a column from the data frame.
* Also, to make permanent changes to the data frame we will have to use `inplace=True` parameter.
* We also see that the index are not correct now as first row has been removed. So, we will have to reset the index of the data frame. Let's see how this can be done.

In [ ]:
# creating a new dataframe
store_data_new  = store_data.drop(1,axis=0)
store_data_new

location gender            type  quantity  total_bill  rating
0        Chicago      M     Electronics         1         100       2
2        Seattle      F  Food&Beverages         4         125       3
3  San Francisco      M        Medicine         2          50       4
4         Austin      F          Beauty         1          80       4

In [ ]:
# resetting the index of data frame
store_data_new.reset_index()

index       location gender            type  quantity  total_bill  rating
0      0        Chicago      M     Electronics         1         100       2
1      2        Seattle      F  Food&Beverages         4         125       3
2      3  San Francisco      M        Medicine         2          50       4
3      4         Austin      F          Beauty         1          80       4

* We see that the index of the data frame is now resetted but the index has become a column in the data frame. We do not need the index to become a column so we can simply set the parameter **`drop=True`** in reset_index() function.

In [ ]:
# setting inplace = True to make the changes permanent
store_data_new.reset_index(drop=True,inplace=True)
store_data_new

location gender            type  quantity  total_bill  rating
0        Chicago      M     Electronics         1         100       2
1        Seattle      F  Food&Beverages         4         125       3
2  San Francisco      M        Medicine         2          50       4
3         Austin      F          Beauty         1          80       4

### 2.8 Pandas - Combining DataFrames

We will examine 3 methods for combining dataframes

1. concat
2. join
3. merge

In [ ]:
data_cust = pd.DataFrame({"customerID":['101','102','103','104'], 
                        'category': ['Medium','Medium','High','Low'],
                        'first_visit': ['yes','no','yes','yes'],
                        'sales': [123,52,214,663]},index=[0,1,2,3])

data_cust_new = pd.DataFrame({"customerID":['101','103','104','105'], 
                    'distance': [12,9,44,21],
                    'sales': [123,214,663,331]},index=[4,5,6,7])

In [ ]:
data_cust

customerID category first_visit  sales
0        101   Medium         yes    123
1        102   Medium          no     52
2        103     High         yes    214
3        104      Low         yes    663

In [ ]:
data_cust_new

customerID  distance  sales
4        101        12    123
5        103         9    214
6        104        44    663
7        105        21    331

In [ ]:
pd.concat([data_cust,data_cust_new],axis=0)

customerID category first_visit  sales  distance
0        101   Medium         yes    123       NaN
1        102   Medium          no     52       NaN
2        103     High         yes    214       NaN
3        104      Low         yes    663       NaN
4        101      NaN         NaN    123      12.0
5        103      NaN         NaN    214       9.0
6        104      NaN         NaN    663      44.0
7        105      NaN         NaN    331      21.0

In [ ]:
pd.concat([data_cust,data_cust_new],axis=1)

customerID category first_visit  sales customerID  distance  sales
0        101   Medium         yes  123.0        NaN       NaN    NaN
1        102   Medium          no   52.0        NaN       NaN    NaN
2        103     High         yes  214.0        NaN       NaN    NaN
3        104      Low         yes  663.0        NaN       NaN    NaN
4        NaN      NaN         NaN    NaN        101      12.0  123.0
5        NaN      NaN         NaN    NaN        103       9.0  214.0
6        NaN      NaN         NaN    NaN        104      44.0  663.0
7        NaN      NaN         NaN    NaN        105      21.0  331.0

**Merge and Join**

* Merge combines dataframes using a column's values to identify common entries

* Join combines dataframes using the index to identify common entries

In [ ]:
pd.merge(data_cust,data_cust_new,how='outer',on='customerID') # outer merge is union of on

customerID category first_visit  sales_x  distance  sales_y
0        101   Medium         yes    123.0      12.0    123.0
1        102   Medium          no     52.0       NaN      NaN
2        103     High         yes    214.0       9.0    214.0
3        104      Low         yes    663.0      44.0    663.0
4        105      NaN         NaN      NaN      21.0    331.0

In [ ]:
pd.merge(data_cust,data_cust_new,how='inner',on='customerID') # inner merge is intersection of on

customerID category first_visit  sales_x  distance  sales_y
0        101   Medium         yes      123        12      123
1        103     High         yes      214         9      214
2        104      Low         yes      663        44      663

In [ ]:
pd.merge(data_cust,data_cust_new,how='right',on='customerID') 

customerID category first_visit  sales_x  distance  sales_y
0        101   Medium         yes    123.0        12      123
1        103     High         yes    214.0         9      214
2        104      Low         yes    663.0        44      663
3        105      NaN         NaN      NaN        21      331

In [ ]:
data_quarters = pd.DataFrame({'Q1': [101,102,103],
                              'Q2': [201,202,203]},
                               index=['I0','I1','I2'])

data_quarters_new = pd.DataFrame({'Q3': [301,302,303],
                                  'Q4': [401,402,403]},
                               index=['I0','I2','I3'])

In [ ]:
data_quarters

Q1   Q2
I0  101  201
I1  102  202
I2  103  203

In [ ]:
data_quarters_new

Q3   Q4
I0  301  401
I2  302  402
I3  303  403

* `join` behaves just like merge,  except instead of using the values of one of the columns to combine data frames, it uses the index labels

In [ ]:
data_quarters.join(data_quarters_new,how='right') # outer, inner, left, and right work the same as merge

Q1     Q2   Q3   Q4
I0  101.0  201.0  301  401
I2  103.0  203.0  302  402
I3    NaN    NaN  303  403

### 2.9 Pandas - Saving and Loading DataFrames

**Note**

In real-life scenario, we deal with much larger datasets that have thousands of rows and multiple columns. It will not be feasible for us to create datasets using multiple lists, especially if the number of columns and rows increases.

So, it is clear we need a more efficient way of handling the data simultaneously at the columns and row levels. In Python, we can import dataset from our local system, from links, or from databases and work on them directly instead of creating our own dataset.

**Loading a CSV file in Python**

**For Jupyter Notebook**
* When the data file and jupyter notebook are in the same folder.

In [ ]:
# Using pd.read_csv() function will work without any path if the notebook and dataset are in the folder

# data = pd.read_csv('StockData.csv')

**For Google Colab with Google Drive**

First, we have to give google colab access to our google drive:

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


Once we have access we can load files from google drive using read_csv() function.

In [ ]:
path="/content/drive/MyDrive/Study/MDS_UTexas/1_PythonFoundations/week2_PythonforDataScience_numpyPandas/StockData.csv" 
data=pd.read_csv(path)
path="/content/drive/MyDrive/Study/MDS_UTexas/1_PythonFoundations/week2_PythonforDataScience_numpyPandas/books.csv" 
books=pd.read_csv(path)
books
books2=books.loc[books['average_rating']>4.5]
books2.loc[books2['ratings_count']<10]

title  \
63                                             鋼之鍊金術師 9   
178                                            鋼之鍊金術師 6   
224                             Stargirl LitPlans on CD   
454   Literature Circle Guide: Bridge to Terabithia:...   
584                       Falling Angels (Luc Actar #1)   
681        Fanning the Flame: Bible  Cross  and Mission   
696                  Tyrannosaurus Wrecks (Stanley  #1)   
741   The John Deere Two-Cylinder Tractor Encycloped...   
757   The American Campaign: U.S. Presidential Campa...   
1837  The Goon Show  Volume 4: My Knees Have Fallen ...   
1856                                          鋼之鍊金術師 10   
1960  Wissenschaft der Logik: Die Lehre Vom Begriff ...   
2244  Getting a Grip on the Basics: Building a Firm ...   
2496                               Come Away My Beloved   
3178  Organizational Architecture: Designs for Chang...   
3438         Bill Gates: Computer Legend (Famous Lives)   
3706  A Wife After God's Own Heart Growth and Study ...   
3833  His Princess Devotional: A Royal Encounter Wit...   
3909                 Little Big Book for God's Children   
3927  Colossians and Philemon: A Critical and Exeget...   
4108                  Willem de Kooning: Late Paintings   
4415                                           Corvette   
4455  The Complete Theory Fun Factory: Music Theory ...   
4917                          Taxation of Mineral Rents   
4933  Oliver Wendell Holmes in Paris: Medicine  Theo...   

                             author  average_rating         isbn13  \
63                   Hiromu Arakawa            4.57  9789861156521   
178                  Hiromu Arakawa            4.58  9789861146140   
224                 Mary B. Collins            4.86  9781583372920   
454                  Tara MacCarthy            5.00  9780439271714   
584               Melissa M. Garcia            4.67  9780595377107   
681                  Chris    Green            5.00  9780310249870   
696                  Laura Driscoll            5.00  9780786845033   
741                   Don Macmillan            4.75  9780760329634   
757               James E. Campbell            5.00  9780890969403   
1837                     NOT A BOOK            5.00  9780563388692   
1856                 Hiromu Arakawa            4.60  9789861156538   
1960  Georg Wilhelm Friedrich Hegel            4.78  9783787307678   
2244                  Beth A. Jones            4.67  9781577948261   
2496             Frances J. Roberts            4.58  9781593109158   
3178                David A. Nadler            4.62  9781555424435   
3438               Sara Barton-Wood            5.00  9780739844328   
3706               Elizabeth George            4.69  9780736911689   
3833            Sheri Rose Shepherd            5.00  9781590529621   
3909                     Alice Wong            4.88  9780941807555   
3927                 R. McL. Wilson            5.00  9780567044716   
4108                Julie Sylvester            5.00  9783829602266   
4415                   Jerry Burton            4.83  9780883631218   
4455              Ian        Martin            5.00  9780851621814   
4917                   Ross Garnaut            5.00  9780198284543   
4933             William C. Dowling            5.00  9781584655800   

     language_code  num_pages  ratings_count  text_reviews_count  \
63             NaN      184.0              4                   0   
178            NaN      191.0              5                   0   
224            eng      210.0              5                   0   
454            eng       32.0              4                   1   
584            eng      216.0              1                   0   
681            eng      336.0              1                   1   
696            eng       24.0              2                   1   
741            eng      256.0              4                   0   
757            eng      314.0              0                   0   
1837          

In [ ]:
flowers = pd.Series([2, 3, 5, 4], index=['lily', 'rose', 'daisy', 'lotus'])
flowers['daisy']

5

In [ ]:
# head() function helps us to see the first 5 rows of the data
data.head()

stock        date    price
0  AAPL  08-02-2013  67.8542
1  AAPL  11-02-2013  68.5614
2  AAPL  12-02-2013  66.8428
3  AAPL  13-02-2013  66.7156
4  AAPL  14-02-2013  66.6556

**Loading an excel file in Python**

In [ ]:
path_excel="/content/drive/MyDrive/Study/MDS_UTexas/1_PythonFoundations/week2_PythonforDataScience_numpyPandas/StockData.xlsx" 
data_excel = pd.read_excel(path_excel)

In [ ]:
data_excel.head()

stock       date    price
0  AAPL 2013-02-08  67.8542
1  AAPL 2013-02-11  68.5614
2  AAPL 2013-02-12  66.8428
3  AAPL 2013-02-13  66.7156
4  AAPL 2013-02-14  66.6556

**Saving a dataset in Python**

**Saving the dataset as a csv file**

To save a dataset as .csv file the syntax used is - 

**data.to_csv('name of the file.csv', index=False)**

In [ ]:
data.to_csv('/content/drive/MyDrive/Study/MDS_UTexas/1_PythonFoundations/week2_PythonforDataScience_numpyPandas/Saved_StockData.csv',index=False)

* In jupyter notebook, the dataset will be saved in the folder where the jupyter notebook is located.
* We can also save the dataset to a desired folder by providing the path/location of the folder.

**Saving the dataset as an excel spreadsheet**

To save a dataset as .xlsx file the syntax used is - 

**data.to_excel('name of the file.xlsx',index=False)**

In [ ]:
data.to_excel('/content/drive/MyDrive/Study/MDS_UTexas/1_PythonFoundations/week2_PythonforDataScience_numpyPandas/Saved_StockData.xlsx',index=False)

### 2.10 Pandas - Functions

**head() - to check the first 5 rows of the dataset**

In [ ]:
data.head()

NameError: ignored

**tail() - to check the last 5 rows of the dataset**

In [ ]:
data.tail()

stock        date  price
5031   ZTS  01-02-2018  77.82
5032   ZTS  02-02-2018  76.78
5033   ZTS  05-02-2018  73.83
5034   ZTS  06-02-2018  73.27
5035   ZTS  07-02-2018  73.86

**shape - to check the number of rows and columns in the dataset**

In [ ]:
data.shape

(5036, 3)

* The dataset has 5036 rows and 3 columns. 

**info() - to check the data type of the columns**

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5036 entries, 0 to 5035
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   stock   5036 non-null   object 
 1   date    5036 non-null   object 
 2   price   5036 non-null   float64
dtypes: float64(1), object(2)
memory usage: 118.2+ KB


* The price column is numeric in nature while the stock and date columns are of object types.

**min() - to check the minimum value of a numeric column**

In [ ]:
data['price'].min()

28.4

**max() - to check the maximum value of a numeric column**

In [ ]:
data['price'].max()

179.26

**unique() - to check the number of unique values that are present in a column**

In [ ]:
data['stock'].unique()

array(['AAPL', 'SNI', 'TJX', 'ZTS'], dtype=object)

**value_counts() - to check the number of values that each unique quantity has in a column**

In [ ]:
data['stock'].value_counts()

AAPL    1259
SNI     1259
TJX     1259
ZTS     1259
Name: stock, dtype: int64

**value_counts(normalize=True) - using the `normalize` parameter and initializing it to True will return the relative frequencies of the unique values.** 

In [ ]:
data['stock'].value_counts(normalize=True)

AAPL    0.25
SNI     0.25
TJX     0.25
ZTS     0.25
Name: stock, dtype: float64

**Statistical Functions**

**mean() - to check the mean (average) value of the column**

In [ ]:
data['price'].mean()

73.05702966640192

**median() - to check the median value of the column**

In [ ]:
data['price'].median()

69.08500000000001

**mode() - to check the mode value of the column**

In [ ]:
data['stock'].mode()

0    AAPL
1     SNI
2     TJX
3     ZTS
dtype: object

**To access a particular mode when the dataset has more than 1 mode**

In [ ]:
#to access the first mode 
data['price'].mode()[0]

74.59

**Group By function**
* Pandas dataframe.groupby() function is used to split the data into groups based on some criteria.

In [ ]:
data.groupby(['stock'])['price'].mean()

stock
AAPL    109.066698
SNI      71.319206
TJX      66.743566
ZTS      45.098648
Name: price, dtype: float64

* Here the groupby function is used to split the data into the 4 stocks that are present in the dataset and then the mean price of each of the 4 stock is calculated.

In [ ]:
# similarly we can get the median price of each stock
data.groupby(['stock'])['price'].median()

* Here the groupby function is used to split the data into the 4 stocks that are present in the dataset and then the median price of each of the 4 stock is calculated.

**Let's create a function to increase the price of the stock by 10%**

In [ ]:
def profit(s):
    return s + s*0.10 # increase of 10%

**The Pandas apply() function lets you to manipulate columns and rows in a DataFrame.**

In [ ]:
data['price'].apply(profit)

* We can now add this updated values in the dataset.

In [ ]:
data['new_price'] =data['price'].apply(profit)
data.head()

**Pandas sort_values() function sorts a data frame in ascending or descending order of passed column.**

In [ ]:
data.sort_values(by='new_price',ascending=False) # by default ascending is set to True

### 2.11 Pandas - Date-time Functions

In [ ]:
import datetime
date = datetime.date(2022, 1, 29)
print(date)

2022-01-29


In [ ]:
mat = np.matrix([[ 1, 4, 9, 121, 144, 169],
                [ 16, 25, 36, 196, 225, 256], 
                [ 49, 64, 81, 289, 324, 361]])
np.mean(mat)

131.66666666666666

In [ ]:
# reading the StockData
path="/content/drive/MyDrive/Python Course/StockData.csv" 
data=pd.read_csv(path)

In [ ]:
import numpy as np
demo_array = np.arange(0,10)
demo_array<3

array([ True,  True,  True, False, False, False, False, False, False,
       False])

In [ ]:
# checking the first 5 rows of the dataset
data.head()

In [ ]:
# checking the data type of columns in the dataset
data.info()

* We observe that the date column is of object type whereas it should be of date time data type.

In [ ]:
# converting the date column to datetime format
data['date']  = pd.to_datetime(data['date'],dayfirst=True)

In [ ]:
data.info()

* We observe that the date column has been converted to datetime format

In [ ]:
data.head()

**The column 'date' is now in datetime format. Now we can change the format of the date to any other format** 

In [ ]:
data['date'].dt.strftime('%m/%d/%Y')

In [ ]:
data['date'].dt.strftime('%m-%d-%y')

**Extracting year from the date column**

In [ ]:
data['date'].dt.year

Creating a new column and adding the extracted year values into the dataframe.

In [ ]:
data['year'] = data['date'].dt.year

**Extracting month from the date column**

In [ ]:
data['date'].dt.month

Creating a new column and adding the extracted month values into the dataframe.

In [ ]:
data['month'] = data['date'].dt.month

**Extracting day from the date column**

In [ ]:
data['date'].dt.day

Creating a new column and adding the extracted day values into the dataframe.

In [ ]:
data['day'] = data['date'].dt.day

In [ ]:
data.head()

* We can see that year, month, and day columns have been added in the dataset.

In [ ]:
# The datetime format is convenient for many tasks!
data['date'][1]-data['date'][0]